### PR TITLE
Switch from SQLTypeInfo to hdk::ir::Type (part 1)

### PR DESCRIPTION
**Pull request by ienkovich**
This PR adds some usages of `hdk::ir::Type` at various places. Mostly for `Datum` functions and `Expr` casting. I made it prior to new types merging to demonstrate how new types' usage might look.

The mix of new and old types gives some inconveniences, like having both dimensions and time units for datetime types, but it will be better and more consistent after a full switch.

Overall, I new type system feels OK. Having separate type IDs for various integer and FP types would allow avoiding nested switch statements, but most probably will increase the number of changes required to add new int/fp types (like int128, fp16). But I don't have a very strong preference here, both options would work.

### CODE REVIEW

**Review by alexbaden — APPROVED** [review]
Overall looks good, I had two comments.

File: Shared/Datum.cpp [hunk 1]
```diff
@@ -443,67 +443,89 @@ size_t hash(Datum datum, const SQLTypeInfo& ti) {
...
       return "f";
-    case kNUMERIC:
-    case kDECIMAL: {
-      double v = (double)d.bigintval / pow(10, ti.get_scale());
-      int size = snprintf(buf, buf_size, "%*.*f", ti.get_dimension(), ti.get_scale(), v);
-      CHECK_LE(0, size) << v << ' ' << ti.to_string();
-      CHECK_LT(size_t(size), buf_size) << v << ' ' << ti.to_string();
+    case hdk::ir::Type::kDecimal: {
+      CHECK_EQ(type->size(), 8);
+      auto precision = type->as<hdk::ir::DecimalType>()->precision();
```
[the remark this answers is not in the document]
Comment (ienkovich) [reply Shared/Datum.cpp:457]: We have a strict mapping of IDs to particular type classes. I even wanted to use static cast in `as` implementation for better performance but left dynamic cast for easier bug catching.

Users can never specify ID when creating type objects, type IDs are determined by protected constructors of derived type classes. Having unit tests to cover all the code in `Type.cpp` and `Context.cpp` would be nice though to make the risk of such nasty bugs minimal.

File: IR/Expr.cpp [hunk 2]
```diff
@@ -123,36 +123,51 @@ int64_t safeScale(T from, unsigned const scale) {
...
-      return constval.doubleval == NULL_DOUBLE;
-    case kNULLT:
-      return constval.bigintval == 0;
-    case kARRAY:
+    case Type::kDate:
+    case Type::kTime:
+    case Type::kTimestamp:
+    case Type::kInterval:
+      return constval.bigintval == NULL_BIGINT;
+    case Type::kFixedLenArray:
```
[the remark this answers is not in the document]
Comment (ienkovich) [reply IR/Expr.cpp:166]: In my view, fixed length and variable arrays are quite different for both internals and users. Types have different attributes, very different encodings, and are handled differently in many cases. I'd keep separate types for them.

**Comment by ienkovich** [comment]
@alexbaden @kurapov-peter Does the approval here mean we are good to go with the new type system? No objections?

**Pull request merged**